lib/valueflow.cpp

```diff
@@ -2331,11 +2331,11 @@ static bool bifurcate(const Token* tok, const std::set<nonneg int>& varids, cons
 }
 
 struct ValueFlowAnalyzer : Analyzer {
-    const TokenList* tokenlist;
+    const TokenList& tokenlist;
```

Comment:
I would prefer to use pointers as members instead of references. Otherwise it makes the class non regular(not assignable or default constructible).

Reply:
Also, when using the object after move, a pointer can be a nullptr which is easier to diagnose the issue.

Reply:
Do we need the functionality? They are local objects that are just being passed around.

The point is to avoid that it could ever be a `nullptr`. Since they are just used internally and the object will always exist it is impossible to happen but I think it would be better not to allow it anyways.

Reply:
> Also, when using the object after move, a pointer can be a nullptr which is easier to diagnose the issue.

That is a good point. But until #4845 is merged they are not being moved at all. As it seems to complicate things and probably won't give any meaningful improvement I am open to dropping it anyways.

Reply:
There is [`cppcoreguidelines-avoid-const-or-ref-data-members`](https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines/avoid-const-or-ref-data-members.html) check in clang tidy which will find these issues. We should enable this check.

Reply:
I don't like that at all. It doesn't just introduce lots of pointers which I am trying to reduce. It will also cause countless objects to become mutable through pointers and that is beyond bad. You need to make more things `const` and not less.

Reply:
We are also using way too many `struct` without any access modifiers allowing modification of any members of them if objects are not longer `const`. We have countless hidden issues because of that which luckily rarely cause issues (as they are mostly in the tests) but just yesterday I ran into one in production code which might cause files to be analyzed with wrong settings.

Reply:
Okay - pointer to `const` objects are still allowed. That makes things just slightly better. But still being able to re-assign them in the class is dangerous. So to properly get a working `const * const`-like construct you would need some data class with accessors.

```cpp
class C1
{
public:
    bool f() const;
};

class C2Data
{
protected:
    C2Data(const C1 &c1) : mC1(&c1) {}

    const C1 &c1() const { return *mC1; }
private:
    const C1 * mC1;
};

class C2 : public C2Data
{
public:
    C2(const C1 &c1) : C2Data(c1) {}
    bool f() {
        return c1().f();
    }
};

static void f()
{
    const C1 c1;
    C2 c2(c1);
}
```

Maybe you have to adjust the inheritance access but if I ever knew something about that I have forgotten it.

Still somebody would be able to do stupid things within `C2Data`. That would require some other check that tells you the data class is only there for storage and accessors.

Reply:
Yea you could use `std::reference_wrapper` here(which is similar to `gsl::non_null`)

Reply:
Alright. I will start testing it after I am done with most of my backlog. That will probably spill into the next dev cycle since I keep unearthing issues addressing other things and I don't want to pile yet another refactoring on top of it.